components/tidb_query/src/codec/mysql/time/datetime.rs

```diff
@@ -0,0 +1,1243 @@
+            }
+        }
+        Ok(())
+    }
+
+    fn parse_frac(input: &[u8], fsp: u8, round: bool) -> Result<(bool, u32)> {
+        // We assume `fsp` is a valid value.
+        assert!(fsp < 7);
+        if input.is_empty() {
+            return Ok((false, 0));
+        }
+
+        let len = input.len();
+        let fsp = usize::from(fsp);
+
+        let (input, len) = if fsp >= len {
+            (input, len)
+        } else {
+            // Read one more digits, needed in round.
+            (&input[..=fsp], fsp + 1)
+        };
+
+        let frac = bytes_to_u32(input)? * TEN_POW[MICRO_WITDH.checked_sub(len).unwrap_or(0)];
+        Ok(if round {
+            // Amplify the frac to 7 digits, more easy to round.
+            let frac = if frac < 1_000_000 { frac * 10 } else { frac };
+
+            let mask = TEN_POW[MICRO_WITDH - fsp];
+            let frac = (frac / mask + 5) / 10 * mask;
+            (frac >= 1_000_000, frac)
+        } else {
+            // Truncate the result if `round` is not enabled
+            // Example:
+            // 1. fsp = 3, frac = 9999000, result = 999
+            // 2. fsp = 6, frac = 9999999, result = 999999
+            let frac = if frac >= 1_000_000 { frac / 10 } else { frac };
+            (false, frac)
+        })
+    }
+
+    fn adjust_year(year: u32) -> u32 {
+        if year <= 69 {
+            2000 + year
+        } else if year >= 70 && year <= 99 {
+            1900 + year
+        } else {
+            year
+        }
+    }
+
+    fn parse_float(
+        ctx: &mut EvalContext,
+        input: &[u8],
+        fsp: u8,
+        time_type: TimeType,
+        round: bool,
+    ) -> Result<Time> {
+        let (whole, frac) = Self::split_frac(input)?;
+        let mut parts = [0u32; 7];
+
+        let year_digits = match whole.len() {
+            14 | 8 => 4,
+            9..=12 | 5..=7 => 2,
+            _ => return Err(Error::truncated()),
+        };
+
+        parts[0] = bytes_to_u32(&whole[..year_digits])?;
+        // If we encounter a two-digit year, translate it to a four-digit year.
+        if year_digits == 2 {
+            parts[0] = Time::adjust_year(parts[0]);
+        }
+        for (i, chunk) in whole[year_digits..].chunks(2).enumerate() {
+            parts[i + 1] = bytes_to_u32(chunk)?;
+        }
+
+        // If we miss the `second`, the fractional part is meaningless.
+        if frac.is_some() && whole.len() != 12 && whole.len() != 14 {
+            return Err(Error::truncated());
+        }
+
+        let (carry, frac) = Self::parse_frac(frac.unwrap_or(&[]), fsp, round)?;
+        parts[6] = frac;
+        if carry {
+            Time::round(&mut parts)?;
+        }
+
+        Self::from_slice(ctx, &parts, fsp, time_type)
+    }
+
+    fn parse_date(input: &[u8]) -> Result<[u32; 3]> {
+        let date = input
+            .split(|x| x.is_ascii_punctuation())
+            .collect::<Vec<_>>();
+        if date.len() != 3 {
+            Err(Error::truncated())
```

Comment:
```
+---------+------+--------------------------------------------------------+
| Warning | 1292 | Incorrect datetime value: '2000,01;01;02     20:20:20' |
+---------+------+-------------------------------------------------------
```

The warnings is very different from MySQL.

I think it is necessary to add `FIXME` in this modules

Reply:
One more thing, you should handle escape character.
```
mysql root@localhost:(none)> select cast('70,01\'01;  ' as datetime);
+----------------------------------+
| cast('70,01\'01;  ' as datetime) |
+----------------------------------+
| 1970-01-01 00:00:00              |
+----------------------------------+
1 row in set
Time: 0.019s
mysql root@localhost:(none)> select cast('70,01'01;  ' as datetime);
(1064, "You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '01' at line 1")
mysql root@localhost:(none)> select cast('70,01.01;  ' as datetime);
+---------------------------------+
| cast('70,01.01;  ' as datetime) |
+---------------------------------+
| 1970-01-01 00:00:00             |
+---------------------------------+
1 row in set
Time: 0.014s
```


And, multiple punctuation character
```
mysql root@localhost:(none)> select cast('70,01\\\\.01;  ' as datetime);
+-------------------------------------+
| cast('70,01\\\\.01;  ' as datetime) |
+-------------------------------------+
| 1970-01-01 00:00:00                 |
+-------------------------------------+
1 row in set
Time: 0.016s
mysql root@localhost:(none)> select cast('70,01....\\\\.01;  ' as datetime);
+-----------------------------------------+
| cast('70,01....\\\\.01;  ' as datetime) |
+-----------------------------------------+
| 1970-01-01 00:00:00                     |
+-----------------------------------------+
1 row in set
Time: 0.014s
mysql root@localhost:(none)> select cast('70,01\;;;;01;  ' as datetime);
+-------------------------------------+
| cast('70,01\;;;;01;  ' as datetime) |
+-------------------------------------+
| 1970-01-01 00:00:00                 |
+-------------------------------------+
1 row in set
Time: 0.006s

```